src/Microsoft.AspNetCore.Mvc.Core/Internal/ProblemDetailsFilter.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.AspNetCore.Mvc.Infrastructure;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class ProblemDetailsFilter : IActionFilter
+    {
+        private readonly IErrorDescriptionFactory _factory;
+
+        public ProblemDetailsFilter(IErrorDescriptionFactory factory)
+        {
+            _factory = factory ?? throw new ArgumentNullException(nameof(factory));
+        }
+
+        public void OnActionExecuting(ActionExecutingContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (context.Result == null || !(context.Result is StatusCodeResult statusCodeResult))
+            {
+                return;
+            }
+
+            var isExactlyStatusCodeResult = context.Result.GetType() == typeof(StatusCodeResult);
+            ProblemDetails problemDetails;
+            switch (context.Result)
+            {
+                case NotFoundResult notFoundResult:
+                    problemDetails = GetNotFoundProblemDetails(notFoundResult);
+                    break;
+                case BadRequestResult badRequestResult:
+                    problemDetails = GetBadRequestProblemDetails(badRequestResult);
+                    break;
+                case UnsupportedMediaTypeResult unsupportedMediaTypeResult:
```

Comment:
This is the only other `StatusCodeResult` subtype that's built in to the system. Seemed low cost to add it.

Reply:
Sure. Might want to comment that we don't really expect user code to return this